hypertrace-graphql-labels-schema/src/main/java/org/hypertrace/graphql/label/joiner/DefaultLabelJoinerBuilder.java

```diff
@@ -0,0 +1,69 @@
+package org.hypertrace.graphql.label.joiner;
+
+import io.reactivex.rxjava3.core.Observable;
+import io.reactivex.rxjava3.core.Single;
+import java.util.Collection;
+import java.util.List;
+import java.util.Map;
+import javax.inject.Inject;
+import lombok.AllArgsConstructor;
+import lombok.extern.slf4j.Slf4j;
+import org.hypertrace.core.graphql.common.request.ContextualRequestBuilder;
+import org.hypertrace.core.graphql.context.GraphQlRequestContext;
+import org.hypertrace.graphql.label.dao.LabelDao;
+import org.hypertrace.graphql.label.dao.LabelResponseConverter;
+import org.hypertrace.graphql.label.schema.Label;
+import org.hypertrace.graphql.label.schema.LabelResultSet;
+
+@Slf4j
+class DefaultLabelJoinerBuilder implements LabelJoinerBuilder {
+
+  private final ContextualRequestBuilder requestBuilder;
+  private final LabelDao labelDao;
+  private final LabelResponseConverter responseConverter;
+
+  @Inject
+  DefaultLabelJoinerBuilder(
+      ContextualRequestBuilder requestBuilder,
+      LabelDao labelDao,
+      LabelResponseConverter responseConverter) {
+    this.requestBuilder = requestBuilder;
+    this.labelDao = labelDao;
+    this.responseConverter = responseConverter;
+  }
+
+  @Override
+  public Single<LabelJoiner> build(GraphQlRequestContext context) {
+    return Single.just(new DefaultLabelJoiner(context));
+  }
+
+  @AllArgsConstructor
+  private class DefaultLabelJoiner implements LabelJoiner {
+    private final GraphQlRequestContext context;
+
+    @Override
+    public <T> Single<Map<T, LabelResultSet>> joinLabels(
+        Collection<T> joinSources, LabelIdGetter<T> labelIdGetter) {
+      return labelDao
+          .getLabels(requestBuilder.build(context))
+          .flatMap(labels -> getLabelResultSetMap(joinSources, labelIdGetter, labels));
+    }
+
+    private <T> Single<Map<T, LabelResultSet>> getLabelResultSetMap(
+        Collection<T> joinSources, LabelIdGetter<T> labelIdGetter, List<Label> labels) {
+      return Observable.fromIterable(joinSources)
+          .flatMapSingle(
+              source ->
+                  getLabelResultSet(source, labelIdGetter, labels)
+                      .map(labelResultSet -> Map.entry(source, labelResultSet)))
```

Comment:
nit: moving this map into the below function allows us to remove the (inline) nesting

Reply:
Did you mean change to this?
```
    private <T> Single<Map<T, LabelResultSet>> getLabelResultSetMap(
        Collection<T> joinSources, LabelIdGetter<T> labelIdGetter, List<Label> labels) {
      return Observable.fromIterable(joinSources)
          .flatMapSingle(source -> getLabelResultSet(source, labelIdGetter, labels))
          .toMap(Map.Entry::getKey, Map.Entry::getValue);
    }

    private <T> Single<Map.Entry<T, LabelResultSet>> getLabelResultSet(
        T source, LabelIdGetter<T> labelIdGetter, List<Label> labels) {
      return labelIdGetter
          .getLabelIds(source)
          .flatMap(labelIds -> responseConverter.convert(labelIds, labels))
          .map(labelResultSet -> Map.entry(source, labelResultSet));
    }
```

Reply:
yeah basically - any way we can avoid the nesting and make it easier to read. Given the new signature though, maybe rename the second method to something like `buildLabelResultSetMapEntry`.

Another aside - you can use `.collect(CollectorUtils.immutableMapEntryCollector());` instead of `.toMap(Map.Entry::getKey, Map.Entry::getValue);` if you find it any clearer (only functional difference is immutability).

Reply:
Done